saw-core/src/Verifier/SAW/Simulator/Prims.hs

```diff
@@ -1232,76 +1235,100 @@ errorOp =
-                                                  y <- g a
-                                                  value x y
-    value VUnit             VUnit             = return VUnit
-    value (VPair x1 x2)     (VPair y1 y2)     = VPair <$> thunk x1 y1 <*> thunk x2 y2
-    value (VRecordValue elems1) (VRecordValue
-                                 (alistAllFields (map fst elems1) ->
-                                  Just elems2)) =
-      VRecordValue <$>
-      zipWithM (\(f,th1) th2 -> (f,) <$> thunk th1 th2) elems1 elems2
-    value (VCtorApp i xv)   (VCtorApp j yv)   | i == j = VCtorApp i <$> thunks xv yv
-    value (VVector xv)      (VVector yv)      = VVector <$> thunks xv yv
-    value (VBool x)         (VBool y)         = VBool <$> bpMuxBool bp b x y
-    value (VWord x)         (VWord y)         = VWord <$> bpMuxWord bp b x y
-    value (VInt x)          (VInt y)          = VInt <$> bpMuxInt bp b x y
-    value (VIntMod n x)     (VIntMod _ y)     = VIntMod n <$> bpMuxInt bp b x y
-    value (VNat m)          (VNat n)          | m == n = return $ VNat m
-    value (VString x)       (VString y)       | x == y = return $ VString x
-    value (VFloat x)        (VFloat y)        | x == y = return $ VFloat x
-    value (VDouble x)       (VDouble y)       | x == y = return $ VDouble y
-    value (VExtra x)        (VExtra y)        = VExtra <$> bpMuxExtra bp b x y
-    value x@(VWord _)       y                 = toVector (bpUnpack bp) x >>= \xv -> value (VVector xv) y
-    value x                 y@(VWord _)       = toVector (bpUnpack bp) y >>= \yv -> value x (VVector yv)
-    value x@(VNat _)        y                 = nat x y
-    value x@(VBVToNat _ _)  y                 = nat x y
-    value x@(VIntToNat _)   y                 = nat x y
-    value (TValue x)        (TValue y)        = TValue <$> tvalue x y
-    value x                 y                 =
+    value :: TValue l -> Value l -> Value l -> MValue l
+    value _ (VNat m)  (VNat n)      | m == n = return $ VNat m
+    value _ (VString x) (VString y) | x == y = return $ VString x
+
+    value (VPiType _ _tp tf) (VFun nm f) (VFun _ g) =
+        return $ VFun nm $ \a ->
+           do tp' <- tf a
+              x <- f a
+              y <- g a
+              value tp' x y
+
+    value VUnitType VUnit VUnit = return VUnit
+    value (VPairType t1 t2) (VPair x1 x2) (VPair y1 y2) =
+      VPair <$> thunk t1 x1 y1 <*> thunk t2 x2 y2
+
+    value (VRecordType fs) (VRecordValue elems1) (VRecordValue elems2) =
+      do let em1 = Map.fromList elems1
+         let em2 = Map.fromList elems2
+         let build (f,tp) = case (Map.lookup f em1, Map.lookup f em2) of
+                              (Just v1, Just v2) ->
+                                 do v <- thunk tp v1 v2
+                                    pure (f,v)
+                              _ -> panic "muxValue" ["Record field missing!", show f]
+         VRecordValue <$> traverse build fs
+
+-- TODO, fix datatypes!
```

Comment:
We need to add this case back in; otherwise it would be a regression

Reply:
CF https://github.com/GaloisInc/saw-script/pull/1298/commits/bbc4549860ed130104784f2cc90f666ba40873ce